### PR TITLE
core, params, tests: add DAO hard-fork balance moves

### PR DESCRIPTION
**Pull request by karalabe**
Prerequisite PRs, merge these first to get rid of duplicate commits:

 - [ ] *core, eth: enforce network split post DAO hard-fork* #2795

---

This PR implements the last component of the DAO hard-fork, namely defines the list of DAO addresses the balances of which should be moved and executes the actual state transformation when the DAO hard-fork block arrives.

The PR also adds the DAO consensus tests from the upstream https://github.com/ethereum/tests repo, although as far as I know those tests are still under development, and hence are subject to change.

*Note, the fork parameters (block number and refund account) are just placeholders, whereas the DAO drain list may or may not change.*

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Mon Jul 11 11:36:03 UTC 2016_

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -813,24 +797,43 @@ func MustMakeChainConfig(ctx *cli.Context) *core.ChainConfig {
...
 	}
-	var homesteadBlockNo *big.Int
-	if ctx.GlobalBool(TestNetFlag.Name) {
-		homesteadBlockNo = params.TestNetHomesteadBlock
-	} else {
-		homesteadBlockNo = params.MainNetHomesteadBlock
+	// Set any missing fields due to them being unset or system upgrade
+	if config.HomesteadBlock == nil {
+		if ctx.GlobalBool(TestNetFlag.Name) {
+			config.HomesteadBlock = new(big.Int).Set(params.TestNetHomesteadBlock)
```
Comment (karalabe) [inline cmd/utils/flags.go]: Pass directly, not copy.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -813,24 +797,43 @@ func MustMakeChainConfig(ctx *cli.Context) *core.ChainConfig {
...
+		} else {
+			config.HomesteadBlock = new(big.Int).Set(params.MainNetHomesteadBlock)
+		}
+	}
+	if config.DAOForkBlock == nil {
+		if ctx.GlobalBool(TestNetFlag.Name) {
+			config.DAOForkBlock = new(big.Int).Set(params.TestNetDAOForkBlock)
+		} else {
+			config.DAOForkBlock = new(big.Int).Set(params.MainNetDAOForkBlock)
+		}
```
Comment (karalabe) [inline cmd/utils/flags.go]: `config.DaoForkSupport = true` (if we want true by default)

File: core/block_validator.go [hunk 3]
```diff
@@ -247,6 +248,33 @@ func ValidateHeader(config *ChainConfig, pow pow.PoW, header *types.Header, pare
...
+// ensure it conforms to hard-fork rules.
+func ValidateHeaderExtraData(config *ChainConfig, header *types.Header) error {
+	// DAO hard-fork extension to the header validity: a) if the node is no-fork,
+	// do not accept blocks in the [fork, fork+10) range with the fork specific
+	// extra-data set; b) if the node is pro-fork, require blocks in the specific
+	// range to have the unique extra-data set.
+	if daoBlock := config.DAOForkBlock; daoBlock != nil {
+		// Check whether the block is among the fork extra-override range
+		limit := new(big.Int).Add(daoBlock, params.DAOForkExtraRange)
+		if daoBlock.Cmp(header.Number) <= 0 && header.Number.Cmp(limit) < 0 {
```
Comment (karalabe) [inline core/block_validator.go]: Flip it... :D

**Comment by aakilfernandes** [comment]
Correct me if I'm wrong, but it looks like this leaves the hard fork block to the user. Does this mean that if different users set different blocks, they'll end up on different forks?

**Comment by karalabe** [comment]
The user can only opt-in/opt-out, default blocks are provided for the
mainnet and testnet. The configurable block number is only meaningful for
private networks (i.e. testing).

**Comment by aakilfernandes** [comment]
Got it, thanks. Has that block number been selected yet?

**Comment by karalabe** [comment]
Superseeded and merged in https://github.com/ethereum/go-ethereum/pull/2814.

**Pull request closed without merging**